src/search/parser_tests.h

```diff
@@ -0,0 +1,210 @@
+        std::cout << indent << "reopen_closed = " << reopen_closed << std::endl;
+        std::cout << indent << "f_eval = " << std::endl;
+        if (f_eval) {
+            f_eval->dump(indent + "    ");
+        } else {
+            std::cout << indent << "    " << "Not specified on command line" << std::endl;
+        }
+        std::cout << indent << "preferred = " << std::endl;
+        for (const auto& h : preferred) {
+            h->dump(indent + "    ");
+        }
+    }
+};
+
+class EagerSearchPlugin : public parser::Plugin<SearchEngine, EagerSearch> {
+public:
+    EagerSearchPlugin()
+        : parser::Plugin<SearchEngine, EagerSearch>("eager", "Eager Search") {;
+        document_synopsis("Eager search synopsis");
+
+        add_option<bool>("reopen_closed", "reopen closed nodes", "false");
+        add_option<std::shared_ptr<Evaluator>>(
+            "f_eval",
+            "set evaluator for jump statistics. "
+            "(Optional; if no evaluator is used, jump statistics will not be displayed.)",
+            parser::ArgumentInfo::NO_DEFAULT);
+        add_list_option<std::shared_ptr<Evaluator>>(
+            "preferred",
+            "use preferred operators of these evaluators", "[]");
+    }
+};
+
+class IteratedSearch : public SearchEngine {
+    std::vector<std::unique_ptr<parser::EnrichedASTNode>> searches;
+public:
+    explicit IteratedSearch(const parser::Options &opts)
+        : SearchEngine(opts),
+          searches() {
+        std::shared_ptr<parser::EnrichedASTNode> searches_AST =
+            opts.get<std::shared_ptr<parser::EnrichedASTNode>>("searches");
+        std::shared_ptr<parser::EnrichedListNode> searches_AST_list =
+                std::dynamic_pointer_cast<parser::EnrichedListNode>(searches_AST);
+        for (const auto & element : searches_AST_list->get_elements()) {
+            searches.push_back(element->clone());
+        }
+    }
+    virtual ~IteratedSearch() = default;
+    void dump(std::string indent) override {
+        std::cout << indent << "IteratedSearch with " << std::endl;
+        indent += "    ";
+        std::cout << indent << "searches = " << std::endl;
+        for (const auto& search_AST : searches) {
+            parser::Any search_any = search_AST->construct();
+            std::shared_ptr<SearchEngine> search = parser::any_cast<std::shared_ptr<SearchEngine>>(search_any);
+            search->dump(indent + "    ");
+        }
+    }
+};
+
+class IteratedSearchPlugin : public parser::Plugin<SearchEngine, IteratedSearch> {
+public:
+    IteratedSearchPlugin()
+        : parser::Plugin<SearchEngine, IteratedSearch>("iterated", "Iterated Search") {;
+        add_option<std::vector<std::shared_ptr<SearchEngine>>>(
+            "searches",
+            "use preferred operators of these evaluators", "[]", parser::Bounds::unlimited(), true);
+    }
+};
+
+class BlindHeuristicPlugin : public parser::Plugin<Evaluator, BlindHeuristic> {
+public:
+    BlindHeuristicPlugin()
```

Comment:
I'm fine with reordering this or creating more virtual methods to have different places to define different parts. Everything has to happen in the constructor, though (or at least be called from it), because the reason we use a class is that we can use static objects to execute code at program initialization time. We also thought about making some of these fields protected and writing directly to it rather than using methods to set the data but the current way is closer to the current code and we wanted to minimize the diff.

Reply:
Do you like this order better? This is the order in the wiki but it mixes documentation and option declaration more than before.

```C++
static class BlindHeuristicPlugin : public parser::Plugin<Evaluator, BlindHeuristic> {
public:
    BlindHeuristicPlugin()
        : parser::Plugin<Evaluator, BlindHeuristic>("blind", "Blind Heuristic") {
        document_synopsis("Returns cost of cheapest action for "
                          "non-goal states, "
                          "0 for goal states");

        add_option<int>("repeat", "how often to check for the goal", "2K");
        add_option<double>("time_limit", "when to stop", "2e4");
        add_list_option<std::vector<int>>("patterns", "...", "[[1,2,3], []]");

        std::vector<std::string> lp_solvers;
        std::vector<std::string> lp_solvers_doc;
        lp_solvers.push_back("CLP");
        lp_solvers_doc.push_back("default LP solver shipped with the COIN library");
        lp_solvers.push_back("CPLEX");
        lp_solvers_doc.push_back("commercial solver by IBM");
        lp_solvers.push_back("GUROBI");
        lp_solvers_doc.push_back("commercial solver");
        lp_solvers.push_back("SOPLEX");
        lp_solvers_doc.push_back("open source solver by ZIB");
        add_enum_option<LPSolverType>(
            "lpsolver",
            lp_solvers,
            "external solver that should be used to solve linear programs",
            "CPLEX",
            lp_solvers_doc);

        document_language_support("action costs", "supported");
        document_language_support("conditional effects", "supported");
        document_language_support("axioms", "supported");

        document_property("admissible", "yes");
        document_property("consistent", "yes");
        document_property("safe", "yes");
        document_property("preferred operators", "no");
    }

    virtual parser::Options preprocess_options(const parser::Options &options) const override {
        options.verify_list_non_empty<std::vector<int>>("patterns");
        parser::Options copy(options);
        std::vector<std::vector<int>> patterns = options.get_list<std::vector<int>>("patterns");
        for (std::vector<int> &pattern : patterns) {
            pattern.push_back(42);
        }
        copy.set("patterns", patterns);
        return copy;
    }
} blind_heuristic_plugin;
```

Reply:
OK, I updated it. It's only in our tests of course, so we have to remember doing this in all plugins when we switch all of them.